### PR TITLE
Make `impAnn` a bit more useful, by making the logs scoped by `impAnn`

### PR DESCRIPTION
**Pull request by lehins**
# Description

Reviewing #3852 made me realize that sometimes we want to display lengthy Imp logs only when there is some failure in the current scope, but not otherwise.

So, imagine we have some very lengthy log entry (in this example: `"Found a problem with x and y"`), but we want to show this only when `x` and `y` are not equal, but for all ther failures in the test this log entry would not be relevant and should not be displayed. Then instead of writing this:

```haskell
when (x /= y) $ do
  logEntry "Found a problem with x and y"
  expectationFailure $ showTreeDiff x y
```

we can now write this:
```haskell
impAnn "Ensuring x and y validity" $ do
  logEntry "Found a problem with x and y"
  x `shouldBeExpr` y
```

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 1]
```diff
@@ -599,14 +599,18 @@ instance Show ImpException where
...
 impAnn :: NFData a => String -> ImpTestM era a -> ImpTestM era a
 impAnn msg m = do
-  catchAnyDeep m $ \exc ->
+  logs <- use impLogL
+  res <- catchAnyDeep m $ \exc ->
     throwIO $
       case fromException exc of
         Just (ImpException ann origExc) -> ImpException (msg : ann) origExc
         Nothing -> ImpException [msg] exc
+  impLogL .= logs
```
Comment (teodanciu) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:612]: But these scopes couldn't be related to each other?  So if we have a failure, there is no chance that the logs in this other scope could help track down the failure?

**Review by teodanciu — APPROVED** [review]
Just a question, approving (since I'm almost certain of the answer, but just to check that I understand). 
Looks good to me

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 1 above)
Comment (Soupstraw) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:612]: I think the idea is that each scope is hand-crafted to help debug a specific problem, so if there is some information you would like to have in multiple scopes, you'll just turn those loggers into a function and run it again in each scope where that information is relevant.

**Review by Soupstraw — APPROVED** [review]
Nice!

**Pull request merged**